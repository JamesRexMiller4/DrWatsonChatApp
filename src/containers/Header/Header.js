import React, { Component } from 'react';
import { connect } from 'react-redux';
import { removeUser, clearMessages} from '../../actions';
import { endConversation } from '../../apiCalls';
import './Header.css'

export class Header extends Component {

  signOut = async () => {
    try {
      await endConversation();
      this.props.removeUser();
      this.props.clearMessages();
      console.log('Working')
    } catch({ message }) {
      this.props.hasErrored(message);
    }
  }

    render () {
      const { user } = this.props
    return (
      <header>
        <h1>Dr. Watson</h1>
        {user && <button onClick={this.signOut}>Sign Out</button>}
      </header>
    )
  }

}
export const mapStateToProps = ({ user, clearMessages }) => ({
  user,
  clearMessages
})

export const mapDispatchToProps = dispatch => ({
  removeUser: () => dispatch( removeUser() ),
  clearMessages: () => dispatch( clearMessages() )
})

export default connect(mapStateToProps, mapDispatchToProps)(Header);