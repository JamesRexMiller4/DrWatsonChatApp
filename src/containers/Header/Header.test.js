import React from 'react';
import { shallow } from 'enzyme';
import { Header, mapStateToProps } from './Header';

describe('Header component', () => {
  let wrapper
  const mockSignOut = jest.fn();
  beforeEach(() => {
    const mockUser = {
      id: 1568665187737, 
      firstName: "Travis", 
      lastName: "Rollins", 
      feeling: "tired"
    };
    wrapper = shallow(<Header user={mockUser} />) 
    wrapper.instance().signOut = jest.fn()
  })
  it('should match the snapshot displaying a logout button if there is a user', () => {
    expect(wrapper).toMatchSnapshot()

  });

  it('should match the snapshot with just the title if there is no user', () => {
    const wrapper = shallow(<Header user={null} signOut={mockSignOut} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call the signOut function when a user clicks on the button', async () => {
    const apiCalls = require('../../apiCalls')

    const endConvoSpy = jest.spyOn(apiCalls, 'endConversation');
    endConvoSpy.mockResolvedValue()

    wrapper.find('button').simulate('click');
    await Promise.resolve()

    expect(endConvoSpy).toHaveBeenCalled();
  })
});

describe('mapStateToProps', () => {
  it('should return an object with the user information', () => {
    const mockUser = {
      id: 1568665187737, 
      firstName: "Travis", 
      lastName: "Rollins", 
      feeling: "tired"
    };

    const mockState = {
      user: mockUser,
      messages: [{
        message: 'Hi there, my name is Dr. Watson. I understand that you have been feeling happy. That is super exciting to hear!',
        isUser: false,
        errorMsg: ''
      }]
    };
    const expected = {
      user: mockUser
    }

    const mappedProps = mapStateToProps(mockState);
    
    expect(mappedProps).toEqual(expected);
  });
});