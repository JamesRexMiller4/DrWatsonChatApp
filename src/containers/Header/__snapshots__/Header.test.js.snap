// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should match the snapshot displaying a logout button if there is a user 1`] = `
<header>
  <h1>
    Dr. Watson
  </h1>
  <button
    onClick={[Function]}
  >
    Sign Out
  </button>
</header>
`;

exports[`Header component should match the snapshot with just the title if there is no user 1`] = `
<header>
  <h1>
    Dr. Watson
  </h1>
</header>
`;
